File: problems/0271-encode-and-decode-strings/typescript/solution.ts
// Encode a list of strings to a single string.
export function encode(strs: string[]): string {
  return strs.map((s) => `${s.length}#${s}`).join("");
}

// Decode a single string back to the list of strings.
export function decode(s: string): string[] {
  const res: string[] = [];
  let i = 0;
  while (i < s.length) {
    let j = i;
    while (s[j] !== "#") j++;
    const length = Number(s.slice(i, j));
    const start = j + 1;
    res.push(s.slice(start, start + length));
    i = start + length;
  }
  return res;
}

File: problems/0271-encode-and-decode-strings/python/solution.py
def encode(strs):
    """Encode a list of strings to a single string."""
    return "".join(f"{len(s)}#{s}" for s in strs)


def decode(s):
    """Decode a single string back to the list of strings."""
    res = []
    i = 0
    while i < len(s):
        j = i
        while s[j] != "#":
            j += 1
        length = int(s[i:j])
        start = j + 1
        res.append(s[start:start + length])
        i = start + length
    return res
